# 🎓 DSLR: Datascience × Logistic Regression
## Harry Potter and a Data Scientist

---

> **Objectif** : Construire un Chapeau de Répartition magique en utilisant la régression logistique et la descente de gradient.  
> **Mission** : Classifier les étudiants de Poudlard dans leurs maisons à partir de leurs notes de cours.

---

Ce notebook explique en détail l'implémentation complète d'un pipeline de machine learning **sans utiliser de bibliothèques ML externes** (pas de pandas, scikit-learn, numpy avec fonctions prêtes à l'emploi).

**Projet développé dans le cadre du cursus IA/ML de 42 Paris**

## 📋 Table des matières

1. [Installation et Configuration](#installation)
2. [Partie 1 : Analyse des Données (describe.py)](#partie1)
3. [Partie 2 : Visualisation des Données](#partie2)
   - 2.1 Histogrammes
   - 2.2 Scatter Plots
   - 2.3 Pair Plot
4. [Partie 3 : Régression Logistique](#partie3)
   - 3.1 Théorie
   - 3.2 Entraînement (logreg_train.py)
   - 3.3 Prédiction (logreg_predict.py)
5. [Résultats et Conclusion](#resultats)

## 🔧 Installation et Configuration {#installation}

Nous commençons par cloner le repository GitHub contenant tout le code du projet.

In [1]:
import os

REPO_URL = "https://github.com/ai-dg/dslr.git"

if not os.path.exists("dslr"):
    print("[>] Cloning repository...")
    get_ipython().system(f"git clone {REPO_URL}")
    print("[+] Repository cloned successfully")
else:
    get_ipython().system("rm -rf ./dslr")
    get_ipython().system(f"git clone {REPO_URL}")
    print("[+] Repository already present, updated")

# Changer de répertoire
os.chdir("dslr")
print(f"[+] Current directory: {os.getcwd()}")

Cloning into 'dslr'...
remote: Enumerating objects: 164, done.
remote: Counting objects: 100% (164/164), done.
remote: Compressing objects: 100% (142/142), done.
remote: Total 164 (delta 59), reused 98 (delta 16), pack-reused 0 (from 0)
Receiving objects: 100% (164/164), 7.71 MiB | 10.50 MiB/s, done.
Resolving deltas: 100% (59/59), done.
[+] Repository already present, updated
[+] Current directory: /home/dagudelo/Projects/dslr/dslr


## 📊 Partie 1 : Analyse des Données (describe.py) {#partie1}

### Objectif
Créer un programme qui analyse statistiquement toutes les colonnes numériques du dataset et affiche des métriques descriptives.

### Métriques calculées manuellement
- **Count** : Nombre d'observations
- **Mean** : Moyenne
- **Std** : Écart-type
- **Min/Max** : Valeurs extrêmes
- **25%, 50%, 75%** : Quartiles
- **Range** : Étendue (Max - Min)
- **Variance** : Variance
- **CV (%)** : Coefficient de variation
- **IQR** : Interquartile Range
- **Low/High IQR** : Bornes pour détecter les outliers
- **Skewness** : Asymétrie de la distribution
- **Kurtosis** : Aplatissement de la distribution

### ⚠️ Contrainte importante
**Toutes ces statistiques doivent être calculées manuellement** - aucune fonction prête à l'emploi n'est autorisée (pas de `.mean()`, `.std()`, etc.)

In [2]:
# Exécution de describe.py
!python3 describe.py datasets/dataset_train.csv

Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0


### Explication du code describe.py

Le fichier `describe.py` implémente plusieurs fonctions clés :

1. **`ft_recovering_data_from_dataset(path)`** : Lit le CSV et convertit les colonnes numériques en float
2. **`ft_recover_numeric_values_from_columns(data)`** : Identifie les colonnes numériques
3. **`ft_recover_stats_from_numeric_cols(data)`** : Calcule toutes les statistiques manuellement
4. **`ft_show_stats_in_terminal(data)`** : Affiche les résultats formatés

**Points techniques importants** :
- Calcul de la variance : $\sigma^2 = \frac{\Sigma(x_i - \mu)^2}{n}$
- Calcul de l'écart-type : $\sigma = \sqrt{\frac{\Sigma(x_i - \mu)^2}{n-1}}$ (correction de Bessel)
- Calcul des quartiles : interpolation linéaire pour les valeurs entre deux points
- Skewness : $\frac{\Sigma(x_i - \mu)^3}{n \times \sigma^3}$
- Kurtosis : $\frac{\Sigma(x_i - \mu)^4}{n \times \sigma^4} - 3$ (kurtosis excess, centré sur 0)

## 📈 Partie 2 : Visualisation des Données {#partie2}

La visualisation est un outil puissant pour comprendre les données et détecter des patterns ou anomalies.

### 2.1 Histogrammes (histogram.py)

**Question à répondre** : Quelle matière de Poudlard a une distribution de scores homogène entre les quatre maisons ?

L'histogramme montre la distribution des scores par maison pour chaque matière.

In [3]:
# Génération des histogrammes
!python3 histogram.py datasets/dataset_train.csv

Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0


### 2.2 Scatter Plots (scatter_plot.py)

**Question à répondre** : Quelles sont les deux features qui sont similaires (forte corrélation) ?

Le scatter plot montre la relation entre deux matières et calcule leur coefficient de corrélation de Pearson.

In [4]:
# Génération des scatter plots (10 paires aléatoires)
!python3 scatter_plot.py datasets/dataset_train.csv

Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0


**Formule de corrélation de Pearson** :
$$r = \frac{\Sigma(x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\Sigma(x_i - \bar{x})^2}\sqrt{\Sigma(y_i - \bar{y})^2}}$$

- **|r| > 0.7** : Corrélation forte
- **0.3 ≤ |r| ≤ 0.7** : Corrélation modérée  
- **|r| < 0.3** : Corrélation faible

### 2.3 Pair Plot (pair_plot.py)

**Question à répondre** : Quelles features allez-vous utiliser pour votre régression logistique ?

Le pair plot affiche une matrice de tous les scatter plots entre toutes les paires de features, permettant d'identifier visuellement les relations et de sélectionner les features les plus pertinentes.

In [5]:
# Génération du pair plot complet
!python3 pair_plot.py datasets/dataset_train.csv

Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0
Non-numeric value found:  replaced by 0


## 🧠 Partie 3 : Régression Logistique {#partie3}

### 3.1 Théorie de la Régression Logistique

La régression logistique est un algorithme de classification qui utilise une fonction sigmoïde pour transformer une combinaison linéaire des features en probabilité.

#### Fonction Hypothèse (Sigmoid)

$$h_\theta(x) = \frac{1}{1 + e^{-\theta^T x}}$$

La sigmoïde transforme n'importe quelle valeur réelle en une probabilité entre 0 et 1.

#### Fonction de Coût (Log-Loss)

$$J(\theta) = -\frac{1}{m} \sum_{i=1}^{m} \Big[ y^{(i)} \log(h_\theta(x^{(i)})) + (1 - y^{(i)}) \log(1 - h_\theta(x^{(i)})) \Big]$$

Cette fonction pénalise fortement les prédictions incorrectes avec une grande confiance.

#### Gradient

$$\frac{\partial J}{\partial \theta_j} = \frac{1}{m} \sum_{i=1}^{m} (h_\theta(x^{(i)}) - y^{(i)}) x_j^{(i)}$$

#### Mise à jour des paramètres

$$\theta_j := \theta_j - \alpha \frac{\partial J}{\partial \theta_j}$$

où $\alpha$ est le taux d'apprentissage (learning rate).

### Classification Multi-classe : One-vs-All

Pour classifier dans 4 maisons, nous entraînons **4 modèles binaires** :
- Modèle 1 : Gryffindor vs (Ravenclaw, Slytherin, Hufflepuff)
- Modèle 2 : Ravenclaw vs (Gryffindor, Slytherin, Hufflepuff)
- Modèle 3 : Slytherin vs (Gryffindor, Ravenclaw, Hufflepuff)
- Modèle 4 : Hufflepuff vs (Gryffindor, Ravenclaw, Slytherin)

Pour prédire, on choisit la maison avec la probabilité la plus élevée parmi les 4 modèles.

### 3.2 Entraînement (logreg_train.py)

Le script `logreg_train.py` entraîne les 4 modèles One-vs-All. Il implémente **3 variantes de descente de gradient** :

1. **BGD (Batch Gradient Descent)** : Utilise tous les exemples à chaque itération
2. **SGD (Stochastic Gradient Descent)** : Met à jour les poids après chaque exemple
3. **Mini-Batch GD** : Utilise des mini-batches de 32 exemples

#### Fonctionnalités bonus implémentées :
- ✅ Normalisation des features (z-score)
- ✅ Export des poids theta en JSON
- ✅ Calcul du log-loss moyen

In [ ]:
# Entraînement avec SGD
# Note: Pour Colab, nous devons simuler l'input
import sys
from io import StringIO

# Sauvegarder stdin original
original_stdin = sys.stdin

# Simuler l'input "SGD"
sys.stdin = StringIO("SGD\n")

try:
    !python3 logreg_train.py datasets/dataset_train.csv
finally:
    # Restaurer stdin
    sys.stdin = original_stdin

Choose training method (BGD, SGD, M-Batch): 

#### Explication du processus d'entraînement

1. **Chargement des données** : Lecture du CSV et extraction des features numériques
2. **Normalisation** : Chaque feature est normalisée : $x_{norm} = \frac{x - \mu}{\sigma}$
3. **Création des labels binaires** : Pour chaque maison, création d'un vecteur binaire (1 si la maison, 0 sinon)
4. **Initialisation des poids** : `theta = [0.0, 0.0, ..., 0.0]` (biais + un poids par feature)
5. **Boucle d'entraînement** :
   - Forward pass : Calcul de $h(x)$ pour tous les exemples
   - Calcul du gradient
   - Mise à jour des poids : $\theta := \theta - \alpha \times gradient$
6. **Sauvegarde** : Les poids sont sauvegardés dans `trained_model.json`

In [ ]:
# Visualisation du modèle entraîné
import json

with open("trained_model.json", "r") as f:
    model = json.load(f)

print("Structure du modèle entraîné :")
print(f"Nombre de maisons : {len(model)}")
print("\nExemple pour Gryffindor :")
print(f"  - Bias (θ₀) : {model['Gryffindor']['theta']['bias']:.6f}")
print(f"  - Nombre de features : {len([k for k in model['Gryffindor']['theta'].keys() if k != 'bias'])}")
print(f"  - Exemples de poids :")
for i, (key, value) in enumerate(list(model['Gryffindor']['theta'].items())[:5]):
    if key != 'bias':
        print(f"    {key}: {value:.6f}")

### 3.3 Prédiction (logreg_predict.py)

Le script `logreg_predict.py` utilise le modèle entraîné pour prédire les maisons des étudiants du dataset de test.

#### Processus de prédiction :

1. **Chargement du modèle** : Lecture de `trained_model.json`
2. **Chargement des données de test** : Lecture du CSV de test
3. **Normalisation** : Application de la même normalisation qu'à l'entraînement
4. **Prédiction** : Pour chaque étudiant :
   - Calcul de la probabilité pour chaque maison
   - Sélection de la maison avec la probabilité maximale
5. **Évaluation** : Calcul de l'accuracy (si les vraies labels sont disponibles)
6. **Export** : Génération de `houses.csv` avec les prédictions

In [ ]:
# Prédiction sur le dataset de test
!python3 logreg_predict.py datasets/dataset_test.csv

#### Visualisation des résultats

In [ ]:
# Affichage des premières prédictions
import csv

print("Premières prédictions :\n")
with open("houses.csv", "r") as f:
    reader = csv.reader(f)
    header = next(reader)
    print(f"{header[0]:<10} {header[1]}")
    print("-" * 30)
    for i, row in enumerate(reader):
        if i < 20:  # Afficher les 20 premières
            print(f"{row[0]:<10} {row[1]}")
        else:
            break

## 📊 Résultats et Conclusion {#resultats}

### Résumé du projet

Ce projet implémente un **pipeline complet de machine learning** sans utiliser de bibliothèques ML externes :

✅ **Analyse statistique complète** : Calcul manuel de toutes les métriques descriptives  
✅ **Visualisations** : Histogrammes, scatter plots, pair plots pour explorer les données  
✅ **Régression logistique One-vs-All** : Classification multi-classe avec 4 modèles binaires  
✅ **Trois méthodes d'optimisation** : BGD, SGD, Mini-Batch GD  
✅ **Normalisation des features** : Améliore la convergence de l'algorithme  
✅ **Évaluation** : Calcul de l'accuracy (objectif : ≥98%)

### Points techniques clés

1. **Tous les calculs sont manuels** : Pas de fonctions prêtes à l'emploi
2. **Gestion des cas limites** : Protection contre overflow dans sigmoid, gestion des valeurs manquantes
3. **Normalisation** : Essentielle pour la convergence avec gradient descent
4. **One-vs-All** : Stratégie efficace pour la classification multi-classe

### Objectifs pédagogiques atteints

- Compréhension approfondie des algorithmes de machine learning
- Implémentation from scratch des concepts fondamentaux
- Analyse et visualisation de données
- Optimisation avec différentes variantes de gradient descent

---

**Projet développé dans le cadre du cursus IA/ML de 42 Paris**  
*Pour des fins éducatives uniquement*

## 📚 Références et Ressources

- **Sujet original** : `en.subject.pdf`
- **Rapport détaillé** : `Report.pdf`
- **Repository GitHub** : https://github.com/ai-dg/dslr.git

### Concepts mathématiques utilisés

- **Régression logistique** : Classification probabiliste
- **Descente de gradient** : Optimisation des paramètres
- **Fonction sigmoïde** : Transformation en probabilité
- **Log-loss** : Fonction de coût pour la classification
- **One-vs-All** : Stratégie de classification multi-classe
- **Normalisation z-score** : Standardisation des features

---

**Fin du notebook** 🎓